src/atomate2/vasp/schemas/magnetism.py

```diff
@@ -0,0 +1,105 @@
+"""Schemas for magnetic ordering calculations in VASP."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from atomate2.common.schemas.magnetism import (
+    MagneticOrderingOutput as MagneticOrderingOutputBase,
+)
+from atomate2.common.schemas.magnetism import (
+    MagneticOrderingRelaxation as MagneticOrderingRelaxationBase,
+)
+from atomate2.common.schemas.magnetism import (
+    MagneticOrderingsDocument as MagneticOrderingsDocumentBase,
+)
+
+if TYPE_CHECKING:
+    from emmet.core.tasks import TaskDoc
+
+
+class MagneticOrderingRelaxation(MagneticOrderingRelaxationBase):
+    """
+    Defines the relaxation output document for a magnetic ordering calculation.
+
+    The construction of this document is implemented here for VASP. See base class for
+    more details.
+    """
+
+    @classmethod
+    def from_task_document(cls, task_document: TaskDoc, uuid: str | None = None):
+        """Construct a MagneticOrderingRelaxation from a task document.
+
+        This does not include the uuid, which must be set separately.
+        """
+        dir_name = task_document.dir_name
+        structure = task_document.structure
+        input_structure = task_document.input.structure
+        if not input_structure.site_properties.get("magmom"):
+            input_structure.add_site_property(  # input struct likely has no magmoms
+                "magmom", task_document.input.parameters["MAGMOM"]
+            )
+        energy = task_document.output.energy
+
+        return cls.from_structures_and_energies(
+            input_structure=input_structure,
+            output_structure=structure,
+            output_energy=energy,
+            uuid=uuid,
+            dir_name=dir_name,
+        )
+
+
+class MagneticOrderingOutput(MagneticOrderingOutputBase):
+    """Defines the static output for a magnetic ordering calculation.
+
+    The construction of this document is implemented here for VASP. See base class for
+    more details.
+    """
+
+    @classmethod
+    def from_task_document(
+        cls,
+        task_document: TaskDoc,
+        uuid: str | None = None,
+        relax_output: MagneticOrderingRelaxation | None = None,
+    ):
+        """Construct a MagneticOrderingOutput from a task document.
+
+        This does not include the uuid, which must be set separately.
+        """
+        dir_name = task_document.dir_name
+        structure = task_document.structure
+        input_structure = task_document.input.structure
+        if not input_structure.site_properties.get("magmom"):
+            input_structure.add_site_property(  # input struct likely has no magmoms
+                "magmom", task_document.input.parameters["MAGMOM"]
+            )
+        energy = task_document.output.energy
```

Comment:
@utf These lines prevent the simpler implementation as you suggest. If magnetic moments were stored in the task document in a universal way (independent of the DFT code), then we would not need to do this, and we could remove the VASP schema subclasses. 

Do you know if there is a code-independent way to access magnetic moments from task documents? They're not always stored in the site properties of the pymatgen structure (hence the logic of the highlighted line).

Reply:
(sorry the highlighted lines should be 74-77)

Reply:
Thanks for the clarification. This is a good point and I'm not sure of the best way forward.

Does it make sense to store the magmoms on the input structure in the task doc? Naively I think it does but perhaps @munrojm or @tschaume can spot some issues.

Alternatively, I propose that we add a new field to the schema, let's say `task_document.input.magnetic_moments` and then make this standard for VASP, CP2K, AIMs etc. 

I think this extra bit of effort will make it much nicer for generalisable workflows longer term.

Reply:
Hi @mattmcdermott, have you had a chance to work on the above at all?

Reply:
@utf No, sorry. It wasn't clear to me which change we should go with and whether I should make the changes in emmet. Are you still wanting to go with the new input field approach?

Reply:
I think the easiest approach for now is to add the new input field. Would you be happy to implement that?

Reply:
@utf Looking into the emmet code further.... if we added a magnetic moments property to the input schema, would this actually be enough? I thought so at first but there is more code here that is VASP specific.

To completely get rid of a DFT code-specific `MagneticOrderingRelaxation.from_task_document` implementation, we would need a universal `TaskDoc` base class that all DFT codes share. As far as I can tell, we don't have this yet. It seems that most task doc implementations for other DFT codes are subclassing `StructureMetadata`, which doesn't necessitate that there is a specific `InputDoc` or `OutputDoc` format, and we also need to ensure that those are in there and follow the same type of structure. Can you confirm that my understanding of this is correct?

If the goal is to write a universal base TaskDoc that all the different DFT-codes share, then I would support that... but it seems like a massive change, and is outside the scope of a single atomate2 feature contribution IMO. (This would likely be something discussed among the full team of developers and implemented as a followup PR right?)

Reply:
Looping in @janosh since he has a better understanding than me of how the subclassing from `emmet-core` works

Reply:
Strictly yes, we should create a universal base TaskDoc, however, I do not expect you do to that as part of this PR. Currently, everything is held together in an ad-hoc way. We don't have a base TaskDoc and yet we do have universal workflows (such as the elastic tensor) due to common fields in the existing task docs. In order to get this merged, I'd be happy with the following:
- Add the requested field for the VASP task doc.
- Raise an issue on atomate2 noting that this field should be added for CP2K/aims.

Reply:
Actually, one is already open from 2022! https://github.com/materialsproject/atomate2/issues/145

Reply:
Oh wow, haha. It agree it should be a priority, and ideally, we can get input from many others to ensure that it does not restrict the addition of support for future codes.

> Add the requested field for the VASP task doc.

And to confirm, once I add the requested field in a PR in `emmet-core`, I can still leave the current VASP-specific implementation for this workflow as is? I think it would have to remain that way unless you want to get rid of the typing (since there is no base TaskDoc we can use to set the type). I guess that would be fine but also seems a little sketchy...

Reply:
Personally, I think it could be better to remove the VASP implementation but add a note in the flow that this workflow currently only works for Task Documents that have this particular field. 

We can remove this note once the universal base is adopted.

Reply:
Alright, got it. I think that will be okay and I'll add a warning.

I'll link to the PR in emmet once it's done.

Reply:
See https://github.com/materialsproject/emmet/pull/945

Reply:
Implemented this change!